File: other/past201912-open/J.nim
import strutils, sequtils, queues


proc solve() =
    var H, W: int
    (H, W) = stdin.readLine.split.map(parseInt)
    let A = mapIt(0..<H, stdin.readLine.split.map(parseInt))
    const
        INF = 1000000000
        dy = [-1, 0, 1, 0]
        dx = [0, 1, 0, -1]
    
    proc bfs(sy, sx: int): int =
        var res: array[50, array[50, int]]
        for y in 0..<H:
            for x in 0..<W:
                res[y][x] = INF
        res[sy][sx] = A[sy][sx]
        var que = initQueue[(int, int)]()
        que.enqueue((sy, sx))
        while que.len != 0:
            let (cy, cx) = que.dequeue
            for k in 0..3:
                let
                    ny = cy + dy[k]
                    nx = cx + dx[k]
                if 0 <= ny and ny < H and 0 <= nx and nx < W:
                    let cost = res[cy][cx] + A[ny][nx]
                    if cost < res[ny][nx]:
                        res[ny][nx] = cost
                        que.enqueue((ny, nx))
        return res[H - 1][0] + res[H - 1][W - 1] + res[0][W - 1] - 2 * res[sy][sx]
    
    var ans = INF
    for i in 0..<H:
        for j in 0..<W:
            ans = min(ans, bfs(i, j))
    echo ans


when is_main_module:
    solve()

File: other/past201912-open/J.py
import sys
from collections import deque

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


def solve():
    H, W = map(int, rl().split())
    A = [list(map(int, rl().split())) for _ in range(H)]
    INF = 10 ** 18
    dy = (-1, 0, 1, 0)
    dx = (0, 1, 0, -1)
    
    def bfs(sy, sx):
        res = [[INF] * W for _ in range(H)]
        res[sy][sx] = 0
        que = deque([[sy, sx]])
        while que:
            cy, cx = que.popleft()
            for k in range(4):
                ny, nx = cy + dy[k], cx + dx[k]
                if 0 <= ny < H and 0 <= nx < W:
                    w = res[cy][cx] + A[ny][nx]
                    if w < res[ny][nx]:
                        res[ny][nx] = w
                        que.append([ny, nx])
        return res
    
    cost1 = bfs(H - 1, 0)
    cost2 = bfs(H - 1, W - 1)
    cost3 = bfs(0, W - 1)
    
    ans = INF
    for i in range(H):
        for j in range(W):
            ans = min(ans, cost1[i][j] + cost2[i][j] + cost3[i][j] - 2 * A[i][j])
    print(ans)


if __name__ == '__main__':
    solve()
